plugins/vpc-tunnel/plugin/commands.go

```diff
@@ -0,0 +1,329 @@
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//     http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package plugin
+
+import (
+	"fmt"
+	"net"
+	"os"
+
+	"github.com/aws/amazon-vpc-cni-plugins/network/eni"
+	"github.com/aws/amazon-vpc-cni-plugins/network/imds"
+	"github.com/aws/amazon-vpc-cni-plugins/network/netns"
+	"github.com/aws/amazon-vpc-cni-plugins/network/utils"
+	tunnelConfig "github.com/aws/amazon-vpc-cni-plugins/plugins/vpc-tunnel/config"
+
+	log "github.com/cihub/seelog"
+	cniSkel "github.com/containernetworking/cni/pkg/skel"
+	cniTypes "github.com/containernetworking/cni/pkg/types"
+	cniTypesCurrent "github.com/containernetworking/cni/pkg/types/current"
+	"github.com/vishvananda/netlink"
+	"golang.org/x/sys/unix"
+)
+
+const (
+	// Name templates used for objects created by this plugin.
+	GeneveLinkNameFormat = "gnv.%s%d"
+	bridgeNameFormat     = "gnvbr.%d"
+
+	// The default MAC address for the geneve interface can be any valid MAC address
+	// since there isn't an actual device as the gateway. The gateway IP address is
+	// just another IP from the same subnet as the primary IP address of the geneve interface.
+	defaultGatewayMacAddr = "00:aa:bb:cc:dd:ee"
+)
+
+// Add is the internal implementation of CNI ADD command.
+func (plugin *Plugin) Add(args *cniSkel.CmdArgs) error {
+	// Parse network configuration.
+	netConfig, err := tunnelConfig.New(args)
+	if err != nil {
+		log.Errorf("Failed to parse netconfig from args: %v.", err)
+		return err
+	}
+
+	log.Infof("Executing ADD with netconfig: %+v.", netConfig)
+
+	// Find the network namespace.
+	log.Infof("Searching for netns %s.", args.Netns)
+	ns, err := netns.GetNetNS(args.Netns)
+	if err != nil {
+		log.Errorf("Failed to find netns %s: %v.", args.Netns, err)
+		return err
+	}
+
+	// Create the Geneve object.
+	geneveName := fmt.Sprintf(GeneveLinkNameFormat, netConfig.VNI, netConfig.DestinationPort)
+	geneve, err := eni.NewGeneve(
+		geneveName,
+		netConfig.VNI,
+		netConfig.DestinationIPAddress,
+		netConfig.DestinationPort,
+		netConfig.Primary)
+	if err != nil {
+		log.Errorf("Failed to create geneve object %s: %v.", geneveName, err)
+		return err
+	}
+
+	// Create an interface for the geneve object.
+	log.Infof("Creating geneve link %s.", geneveName)
+	err = geneve.AttachToLink()
+	if err != nil {
+		if err.Error() == "device or resource busy" {
+			// If the geneve link already exists, it may have been created in a previous invocation
+			// of this plugin. Look for it in the target network namespace and reset it.
+			err = ns.Run(func() error {
+				err := geneve.ENI.AttachToLink()
```

Comment:
Is it possible to refactor this block into a named method? Right now, we get into 4 and 5 levels of indentation down and that makes it hard to read the code. Besides the `Add()` method is too long imho.

Reply:
Refactored in following revision.